File: apps/android/app/src/main/kotlin/ai/parkshield/android/core/network/ApiClient.kt
package ai.parkshield.android.core.network

import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedOutputStream
import java.net.HttpURLConnection
import java.net.URL

/** Feature-owned adapters provide endpoint-specific requests in later vertical slices. */
interface ApiClient {
    suspend fun execute(request: ApiRequest): ApiResult
}

data class ApiRequest(
    val method: String,
    val path: String,
    val body: ByteArray? = null,
    val headers: Map<String, String> = emptyMap(),
)

sealed interface ApiResult {
    data class Success(val statusCode: Int, val body: ByteArray) : ApiResult
    data class Failure(val statusCode: Int, val error: ApiError?) : ApiResult
}

class HttpUrlConnectionApiClient(private val baseUrl: URL) : ApiClient {
    override suspend fun execute(request: ApiRequest): ApiResult {
        val url = URL(baseUrl, request.path.removePrefix("/"))
        val connection = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = request.method
            connectTimeout = TIMEOUT_MS
            readTimeout = TIMEOUT_MS
            setRequestProperty("Accept", "application/json")
            request.headers.forEach(::setRequestProperty)
            request.body?.let { payload ->
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
                BufferedOutputStream(outputStream).use { it.write(payload) }
            }
        }
        return try {
            val status = connection.responseCode
            val body = (if (status in 200..299) connection.inputStream else connection.errorStream)
                ?.use { it.readBytes() } ?: ByteArray(0)
            if (status in 200..299) ApiResult.Success(status, body)
            else ApiResult.Failure(status, parseError(body))
        } finally {
            connection.disconnect()
        }
    }

    private fun parseError(body: ByteArray): ApiError? = try {
        val json = JSONObject(body.decodeToString())
        val details = json.optJSONArray("details")?.toDetails() ?: emptyList()
        ApiError(
            version = json.getString("version"), code = json.getString("code"),
            message = json.getString("message"), correlationId = json.getString("correlation_id"), details = details,
        ).takeIf(ApiErrorContract::isValid)
    } catch (_: Exception) { null }

    private fun JSONArray.toDetails(): List<ApiErrorDetail> = buildList {
        for (index in 0 until length()) {
            val value = getJSONObject(index)
            add(ApiErrorDetail(value.getString("field"), value.getString("code")))
        }
    }

    private companion object { const val TIMEOUT_MS = 15_000 }
}

File: apps/ios/Sources/ParkShieldFoundation/Core/Network/APIClient.swift
import Foundation

public protocol APIClient: Sendable {
    func execute(_ request: APIRequest) async throws -> APIResult
}

public struct APIRequest: Sendable {
    let method: String
    let path: String
    let body: Data?
    let headers: [String: String]

    public init(method: String, path: String, body: Data? = nil, headers: [String: String] = [:]) {
        self.method = method; self.path = path; self.body = body; self.headers = headers
    }
}

public enum APITransportError: Error, Sendable { case invalidURL, requestFailed }

public struct URLSessionAPIClient: APIClient {
    let baseURL: URL
    let session: URLSession
    public init(baseURL: URL, session: URLSession = .shared) { self.baseURL = baseURL; self.session = session }

    public func execute(_ request: APIRequest) async throws -> APIResult {
        guard let url = URL(string: request.path.trimmingCharacters(in: CharacterSet(charactersIn: "/")), relativeTo: baseURL) else { throw APITransportError.invalidURL }
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = request.method; urlRequest.httpBody = request.body; urlRequest.timeoutInterval = 15
        urlRequest.setValue("application/json", forHTTPHeaderField: "Accept")
        if request.body != nil { urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type") }
        request.headers.forEach { urlRequest.setValue($0.value, forHTTPHeaderField: $0.key) }
        let (data, response) = try await session.data(for: urlRequest)
        guard let http = response as? HTTPURLResponse else { throw APITransportError.requestFailed }
        if (200..<300).contains(http.statusCode) { return .success(statusCode: http.statusCode, body: data) }
        let error = try? JSONDecoder().decode(APIError.self, from: data)
        return .failure(statusCode: http.statusCode, error: error.flatMap { APIErrorContract.isValid($0) ? $0 : nil })
    }
}

public enum APIResult: Sendable {
    case success(statusCode: Int, body: Data)
    case failure(statusCode: Int, error: APIError?)
}
